include/ginkgo/extensions/kokkos/types.hpp

```diff
@@ -0,0 +1,288 @@
+// SPDX-FileCopyrightText: 2017-2023 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GINKGO_EXTENSIONS_KOKKOS_TYPES_HPP
+#define GINKGO_EXTENSIONS_KOKKOS_TYPES_HPP
+
+#include <ginkgo/config.hpp>
+
+#if GINKGO_EXTENSION_KOKKOS
+
+#include <Kokkos_Complex.hpp>
+#include <Kokkos_Core.hpp>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+namespace ext {
+namespace kokkos {
+namespace detail {
+
+
+/**
+ * Maps arithmetic types to corresponding Kokkos types.
+ *
+ * @tparam T  An arithmetic type.
+ */
+template <typename T>
+struct value_type_impl {
+    using type = T;
+};
+
+template <typename T>
+struct value_type_impl<const T> {
+    using type = const typename value_type_impl<T>::type;
+};
+
+template <typename T>
+struct value_type_impl<std::complex<T>> {
+    using type = Kokkos::complex<T>;
+};
+
+
+template <typename T>
+struct value_type {
+    using type = typename value_type_impl<T>::type;
+
+    static_assert(sizeof(std::decay_t<T>) == sizeof(std::decay_t<type>),
+                  "Can't handle C++ data type and corresponding Kokkos "
+                  "type with mismatching type sizes.");
+    static_assert(
+        alignof(std::decay_t<T>) == alignof(std::decay_t<type>),
+        "Can't handle C++ data type and corresponding Kokkos type with "
+        "mismatching alignments. If std::complex is used, please make sure "
+        "to configure Kokkos with `KOKKOS_ENABLE_COMPLEX_ALIGN=ON`");
+};
+
+template <typename T>
+using value_type_t = typename value_type<T>::type;
+
+
+template <typename T, typename MemorySpace>
+struct mapper {
+    static auto map(T&);
+    static auto map(const T&);
```

Comment:
I don't think it's possible to document these functions, because they depend too much on the specialization. I've documented the class itself, maybe that's enough.